### ISSUE TITLE
ICapTransaction on CapEFDbTransaction is private.

### ISSUE THREAD

**Issue by thisispaulsmith** [dotnetcore/CAP#867]
It's currently not possible to access `ICapTransaction` from `CapEFDbTransaction` which makes Dapper integration difficult as you can't access the underlying transaction. 

Any objection to a PR to make this public?

**Comment by thisispaulsmith** [comment]
Actually a better solution would be, just to make `CapEFDbTransaction` implement the correct interface. Then you can always reference the transaction using the extension method MS provides.

```csharp
internal class CapEFDbTransaction : IDbContextTransaction, IInfrastructure<DbTransaction>
{
    private readonly ICapTransaction _transaction;

    public CapEFDbTransaction(ICapTransaction transaction)
    {
        _transaction = transaction;
        var dbContextTransaction = (IDbContextTransaction) _transaction.DbTransaction;
        TransactionId = dbContextTransaction.TransactionId;
    }

    public Guid TransactionId { get; }

    public void Dispose()
    {
        _transaction.Dispose();
    }

    public void Commit()
    {
        _transaction.Commit();
    }

    public void Rollback()
    {
        _transaction.Rollback();
    }

    public async Task CommitAsync(CancellationToken cancellationToken = default)
    {
        await _transaction.CommitAsync(cancellationToken);
    }

    public async Task RollbackAsync(CancellationToken cancellationToken = default)
    {
        await _transaction.RollbackAsync(cancellationToken);
    }

    public ValueTask DisposeAsync()
    {
        return new ValueTask(Task.Run(() => _transaction.Dispose()));
    }

    public DbTransaction Instance
    {
        get
        {
            var dbContextTransaction = (IDbContextTransaction)_transaction.DbTransaction;
            return dbContextTransaction.GetDbTransaction();
        }
    }
}
```

**Comment by thisispaulsmith** [comment]
This is the scenario. Just to add a dapper command. Without the above it's not possible unless you introduce some nasty reflection.

```csharp
using (var trans = _context.Database.BeginTransaction(_capBus, autoCommit: false))
{
    _context.Persons.Add(new Person() { Name = "ef.transaction" });

    _context.Database.GetDbConnection()
        .QueryAsync("somesql", transaction: trans.GetDbTransaction());
    
    for (int i = 0; i < 1; i++)
    {
        _capBus.Publish("sample.rabbitmq.mysql", DateTime.Now);
    }

    _context.SaveChanges();

    trans.Commit();
}
```

**Comment by yang-xiaodong** [comment]
Hello @thisispaulsmith ,

Can you explain it? I did not understand the role of the `IInfrastructure<DbTransaction>` interface in the following code?

```
  _context.Database.GetDbConnection().QueryAsync("somesql", transaction: trans.GetDbTransaction());
```

I know that `GetDbTransaction` is an extension method, but it comes from EntityFrameworkCore

```C#
using System.Data.Common;

namespace Microsoft.EntityFrameworkCore.Storage
{
    //
    // Summary:
    //     Extension methods for Microsoft.EntityFrameworkCore.Storage.IDbContextTransaction.
    public static class DbContextTransactionExtensions
    {
        //
        // Summary:
        //     Gets the underlying System.Data.Common.DbTransaction for the given transaction.
        //     Throws if the database being targeted is not a relational database that uses
        //     System.Data.Common.DbTransaction.
        //
        // Parameters:
        //   dbContextTransaction:
        //     The transaction to get the System.Data.Common.DbTransaction from.
        //
        // Returns:
        //     The underlying System.Data.Common.DbTransaction.
        public static DbTransaction GetDbTransaction([NotNullAttribute] this IDbContextTransaction dbContextTransaction);
    }
}
```

Do you mean that if we don't implement  `IInfrastructure<DbTransaction>`, we will get an exception when calling `GetDbTransaction`?

**Comment by thisispaulsmith** [comment]
Yes. You will get an error if the interface is not implemented. The MS extension method looks like this

```csharp
public static DbTransaction GetDbTransaction(this IDbContextTransaction dbContextTransaction)
        {
            Check.NotNull(dbContextTransaction, nameof(dbContextTransaction));

            if (!(dbContextTransaction is IInfrastructure<DbTransaction> accessor))
            {
                throw new InvalidOperationException(RelationalStrings.RelationalNotInUse);
            }

            return accessor.GetInfrastructure();
        }
```

https://github.com/dotnet/efcore/blob/main/src/EFCore.Relational/Storage/DbContextTransactionExtensions.cs

**Comment by thisispaulsmith** [comment]
@yang-xiaodong chance of a preview release?

**Comment by yang-xiaodong** [comment]
Fixed in v5.0.3, NuGet release was completed a few minutes ago
